src/jit/hwintrinsiccodegenxarch.cpp

```diff
@@ -2277,52 +2302,55 @@ void CodeGen::genBMI1Intrinsic(GenTreeHWIntrinsic* node)
             break;
         }
 
-        default:
+        case NI_BMI2_MultiplyNoFlags:
+        case NI_BMI2_X64_MultiplyNoFlags:
         {
-            unreached();
-            break;
-        }
-    }
-
-    genProduceReg(node);
-}
-
-//------------------------------------------------------------------------
-// genBMI2Intrinsic: Generates the code for a BMI2 hardware intrinsic node
-//
-// Arguments:
-//    node - The hardware intrinsic node
-//
-void CodeGen::genBMI2Intrinsic(GenTreeHWIntrinsic* node)
-{
-    NamedIntrinsic intrinsicId = node->gtHWIntrinsicId;
-    regNumber      targetReg   = node->gtRegNum;
-    GenTree*       op1         = node->gtGetOp1();
-    GenTree*       op2         = node->gtGetOp2();
-    var_types      baseType    = node->gtSIMDBaseType;
-    var_types      targetType  = node->TypeGet();
-    instruction    ins         = HWIntrinsicInfo::lookupIns(intrinsicId, targetType);
-    emitter*       emit        = getEmitter();
+            int numArgs = HWIntrinsicInfo::lookupNumArgs(node);
+            assert(numArgs == 2 || numArgs == 3);
+            regNumber op1Reg = REG_NA;
+            regNumber op2Reg = REG_NA;
+            regNumber op3Reg = REG_NA;
+            regNumber lowReg = REG_NA;
+            if (numArgs == 2)
+            {
+                op1Reg = op1->gtRegNum;
+                op2Reg = op2->gtRegNum;
+                lowReg = targetReg;
+            }
+            else
+            {
+                GenTreeArgList* argList = op1->AsArgList();
+                op1                     = argList->Current();
+                genConsumeRegs(op1);
+                op1Reg  = op1->gtRegNum;
+                argList = argList->Rest();
+                op2     = argList->Current();
+                genConsumeRegs(op2);
+                op2Reg       = op2->gtRegNum;
+                argList      = argList->Rest();
+                GenTree* op3 = argList->Current();
+                genConsumeRegs(op3);
+                op3Reg = op3->gtRegNum;
+                lowReg = node->GetSingleTempReg();
+                assert(lowReg != targetReg);
+            }
+            emitAttr attr = emitTypeSize(targetType);
+            // mov the first operand into implicit source operand EDX/RDX
+            if (op1Reg != REG_EDX)
+            {
+                assert(op2Reg != REG_EDX);
```

Comment:
You should also check `op3Reg`, which you can do unconditionally, since it will either be a valid reg or `REG_NA`, and either way it must not be the same as `op1Reg`.

Reply:
Thank you, fixed.

Reply:
I don't see the assert for `op3Reg`

Reply:
@CarolEidt The asserts for op3Reg are in the 3-arg path 😄 
https://github.com/dotnet/coreclr/pull/21480/files#diff-847442c7efdf7a78e78711bea30fd4b2R2336

Reply:
Ah, I see. Thanks.